server/reload.go

```diff
@@ -1006,26 +1015,56 @@ func (s *Server) diffOptions(newOpts *Options) ([]option, error) {
 			if new != old {
 				diffOpts = append(diffOpts, &jetStreamOption{newValue: new})
 			}
+
+			// Mark whether JS will be disabled.
+			disableJS = !new
 		case "storedir":
 			new := newValue.(string)
 			old := oldValue.(string)
-			// If they are the same, or, we have turned it off (-1) that is ok.
-			if new != _EMPTY_ && new != old {
-				return nil, fmt.Errorf("config reload not supported for jetstream storage directory")
-			}
-		case "jetstreammaxmemory":
-			old := oldValue.(int64)
-			new := newValue.(int64)
-			// If they are the same, or, we have turned it off (-1) that is ok.
-			if new != -1 && new != old {
-				return nil, fmt.Errorf("config reload not supported for jetstream max memory")
+			modified := new != old
+
+			// Check whether JS is being disabled and/or storage dir attempted to change.
+			if jsEnabled && modified {
+				if new == _EMPTY_ {
+					// This means that either JS is being disabled or it is using an temp dir.
+					// Allow the change but error in case JS was not disabled.
+					jsStoreDirChanged = true
+				} else {
+					return nil, fmt.Errorf("config reload not supported for jetstream storage directory")
+				}
 			}
-		case "jetstreammaxstore":
+		case "jetstreammaxmemory", "jetstreammaxstore":
 			old := oldValue.(int64)
 			new := newValue.(int64)
-			// If they are the same, or, we have turned it off (-1) that is ok.
-			if new != -1 && new != old {
-				return nil, fmt.Errorf("config reload not supported for jetstream max storage")
+
+			// Check whether JS is being disabled and/or limits are being changed.
```

Comment:
Should this be triggered here?

Reply:
Cannot trigger the error here since it would prevent being able to disable JS, so wait until the end of processing the options after knowing whether JS will be disabled or not to error in case the limits changed.